In [113]:
from IPython import display
import requests
import json
from sodapy import Socrata
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import warnings
import time
warnings.filterwarnings("ignore")
pd.set_option('display.max_colwidth',100)
pd.set_option('display.max_columns',50)

from sklearn.preprocessing import OneHotEncoder


# Introduction

According to the CDC, more than 48 million Americans per year become sick from food, and an estimated 75% of the outbreaks came from food prepared by caterers, delis, and restaurants. In most cities, health inspections are generally random, which can increase time spent on spot checks at clean restaurants that have been following the rules closely — and missed opportunities to improve health and hygiene at places with more pressing food safety issues.

The goal for this project is to leverage public citizen generated data from social media to narrow the search for critical health and safety violations in New York City. As the City of New York manages  an open data portal, everyone can access historical hygiene inspections and violation records. By combine these two data source this project aims to determine which words, phrases, ratings, and patterns among restaurants lead to critical health and safety violations. This model can assist city health inspectors do their job better by prioritizing the kitchens most likely to be in violation of code.

# Obtain

This project requires data pulled from two different sources, the City of New York and Yelp. To obtain the data we will call the API keys.

## NYC Open Data Portal

In [114]:
doh = pd.read_csv('data/DOHMH_New_York_City_Restaurant_Inspection_Results.csv')
doh

CAMIS                                 DBA       BORO BUILDING  \
0       50008319                            POULETTE  Manhattan      790   
1       41678734                        TINY'S DINER      Bronx     3603   
2       40795021                      TAO RESTAURANT  Manhattan       42   
3       40795021                      TAO RESTAURANT  Manhattan       42   
4       40538234            DUNKIN',' BASKIN ROBBINS     Queens     9925   
...          ...                                 ...        ...      ...   
186222  41556790  FIVE GUYS FAMOUS BURGERS AND FRIES  Manhattan     2847   
186223  50032876                      TANNER SMITH'S  Manhattan      204   
186224  50003842                               T- 45  Manhattan      135   
186225  41640824           RICURAS ECUADORIAN BAKERY      Bronx     1576   
186226  41450971               GIOVANNI'S RESTAURANT      Bronx     3209   

                           STREET  ZIPCODE       PHONE CUISINE DESCRIPTION  \
0                        9 AVENUE  10019.0  2129569488             Chicken   
1                RIVERDALE AVENUE  10463.0  7187087600            American   
2                EAST   58 STREET  10022.0  2128882288  Asian/Asian Fusion   
3                EAST   58 STREET  10022.0  2128882288  Asian/Asian Fusion   
4       HORACE HARDING EXPRESSWAY  11368.0  7182719222              Donuts   
...                           ...      ...         ...                 ...   
186222                  BROADFWAY      NaN  2126787701          Hamburgers   
186223           WEST   55 STREET  10019.0  9175172283            American   
186224           WEST   45 STREET  10036.0  6466403775            American   
186225              WATSON AVENUE  10472.0  7184508363             Spanish   
186226         WESTCHESTER AVENUE  10461.0  7188924340               Pizza   

       INSPECTION DATE                                           ACTION  \
0           01/28/2019  Violations were cited in the following area(s).   
1           09/23/2019  Violations were cited in the following area(s).   
2           03/11/2020  Violations were cited in the following area(s).   
3           03/11/2020  Violations were cited in the following area(s).   
4           07/29/2021  Violations were cited in the following area(s).   
...                ...                                              ...   
186222      03/01/2019  Violations were cited in the following area(s).   
186223      06/07/2019  Violations were cited in the following area(s).   
186224      07/26/2018  Violations were cited in the following area(s).   
186225      07/20/2018  Violations were cited in the following area(s).   
186226      06/21/2017  Violations were cited in the following area(s).   

       VIOLATION CODE  \
0                 04D   
1                 04M   
2                 10F   
3                 10F   
4                 06E   
...               ...   
186222            10F   
186223            08A   
186224            10B   
186225            06F   
186226            06D   

                                                                                      VIOLATION DESCRIPTION  \
0       Food worker does not wash hands thoroughly after using the toilet, coughing, sneezing, smoking, ...   
1                                            Live roaches present in facility's food and/or non-food areas.   
2       Non-food contact surface improperly constructed. Unacceptable material used. Non-food contact su...   
3       Non-food contact surface improperly constructed. Unacceptable material used. Non-food contact su...   
4       Sanitized equipment or utensil, including in-use food dispensing utensil, improperly used or sto...   
...                                                                                                     ...   
186222  Non-food contact surface improperly constructed. Unacceptable material used. Non-food contact su...   
186223  Facility not vermin proof. Harborage or conditions conducive to

In [115]:
doh.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186227 entries, 0 to 186226
Data columns (total 26 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   CAMIS                  186227 non-null  int64  
 1   DBA                    186200 non-null  object 
 2   BORO                   186227 non-null  object 
 3   BUILDING               185799 non-null  object 
 4   STREET                 186227 non-null  object 
 5   ZIPCODE                183146 non-null  float64
 6   PHONE                  186214 non-null  object 
 7   CUISINE DESCRIPTION    186226 non-null  object 
 8   INSPECTION DATE        186227 non-null  object 
 9   ACTION                 186227 non-null  object 
 10  VIOLATION CODE         185413 non-null  object 
 11  VIOLATION DESCRIPTION  185863 non-null  object 
 12  CRITICAL FLAG          186227 non-null  object 
 13  SCORE                  186227 non-null  int64  
 14  GRADE                  67933 non-nul

In [123]:
doh.dropna(subset=['PHONE'],inplace=True)

In [124]:
doh.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 186214 entries, 0 to 186226
Data columns (total 26 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   CAMIS                  186214 non-null  int64  
 1   DBA                    186187 non-null  object 
 2   BORO                   186214 non-null  object 
 3   BUILDING               185786 non-null  object 
 4   STREET                 186214 non-null  object 
 5   ZIPCODE                183133 non-null  float64
 6   PHONE                  186214 non-null  object 
 7   CUISINE DESCRIPTION    186213 non-null  object 
 8   INSPECTION DATE        186214 non-null  object 
 9   ACTION                 186214 non-null  object 
 10  VIOLATION CODE         185400 non-null  object 
 11  VIOLATION DESCRIPTION  185850 non-null  object 
 12  CRITICAL FLAG          186214 non-null  object 
 13  SCORE                  186214 non-null  int64  
 14  GRADE                  67924 non-nul

In [125]:
# How many unique restaurants are in this dataset?
doh['CAMIS'].nunique()

19790

Health code violations found during an inspections carries a point value, and a restaurant’s score corresponds to a letter grade. A lower point score, leads to a better letter grade:

"A" grade: 0 to 13 points for sanitary violations
"B" grade: 14 to 27 points for sanitary violations
"C" grade: 28 or more points for sanitary violations

In [126]:
doh['A'] = doh['SCORE'] < 14
doh['B'] = (doh['SCORE'] > 13) & (doh['SCORE'] < 28)
doh['C'] = doh['SCORE'] > 27

In [127]:
doh

CAMIS                                 DBA       BORO BUILDING  \
0       50008319                            POULETTE  Manhattan      790   
1       41678734                        TINY'S DINER      Bronx     3603   
2       40795021                      TAO RESTAURANT  Manhattan       42   
3       40795021                      TAO RESTAURANT  Manhattan       42   
4       40538234            DUNKIN',' BASKIN ROBBINS     Queens     9925   
...          ...                                 ...        ...      ...   
186222  41556790  FIVE GUYS FAMOUS BURGERS AND FRIES  Manhattan     2847   
186223  50032876                      TANNER SMITH'S  Manhattan      204   
186224  50003842                               T- 45  Manhattan      135   
186225  41640824           RICURAS ECUADORIAN BAKERY      Bronx     1576   
186226  41450971               GIOVANNI'S RESTAURANT      Bronx     3209   

                           STREET  ZIPCODE       PHONE CUISINE DESCRIPTION  \
0                        9 AVENUE  10019.0  2129569488             Chicken   
1                RIVERDALE AVENUE  10463.0  7187087600            American   
2                EAST   58 STREET  10022.0  2128882288  Asian/Asian Fusion   
3                EAST   58 STREET  10022.0  2128882288  Asian/Asian Fusion   
4       HORACE HARDING EXPRESSWAY  11368.0  7182719222              Donuts   
...                           ...      ...         ...                 ...   
186222                  BROADFWAY      NaN  2126787701          Hamburgers   
186223           WEST   55 STREET  10019.0  9175172283            American   
186224           WEST   45 STREET  10036.0  6466403775            American   
186225              WATSON AVENUE  10472.0  7184508363             Spanish   
186226         WESTCHESTER AVENUE  10461.0  7188924340               Pizza   

       INSPECTION DATE                                           ACTION  \
0           01/28/2019  Violations were cited in the following area(s).   
1           09/23/2019  Violations were cited in the following area(s).   
2           03/11/2020  Violations were cited in the following area(s).   
3           03/11/2020  Violations were cited in the following area(s).   
4           07/29/2021  Violations were cited in the following area(s).   
...                ...                                              ...   
186222      03/01/2019  Violations were cited in the following area(s).   
186223      06/07/2019  Violations were cited in the following area(s).   
186224      07/26/2018  Violations were cited in the following area(s).   
186225      07/20/2018  Violations were cited in the following area(s).   
186226      06/21/2017  Violations were cited in the following area(s).   

       VIOLATION CODE  \
0                 04D   
1                 04M   
2                 10F   
3                 10F   
4                 06E   
...               ...   
186222            10F   
186223            08A   
186224            10B   
186225            06F   
186226            06D   

                                                                                      VIOLATION DESCRIPTION  \
0       Food worker does not wash hands thoroughly after using the toilet, coughing, sneezing, smoking, ...   
1                                            Live roaches present in facility's food and/or non-food areas.   
2       Non-food contact surface improperly constructed. Unacceptable material used. Non-food contact su...   
3       Non-food contact surface improperly constructed. Unacceptable material used. Non-food contact su...   
4       Sanitized equipment or utensil, including in-use food dispensing utensil, improperly used or sto...   
...                                                                                                     ...   
186222  Non-food contact surface improperly constructed. Unacceptable material used. Non-food contact su...   
186223  Facility not vermin proof. Harborage or conditions conducive to

In [128]:
doh['PHONE'] = '+1'+doh['PHONE']

Of the 19,792 unique restaurants, 12,221 did not pass the intitial inspections at least once.

In [129]:
doh_grades = doh.groupby(['CAMIS'])['PHONE','BORO','BUILDING','STREET','ZIPCODE',
                                    'PHONE','CUISINE DESCRIPTION',
                                    'Latitude','Longitude','Community Board',
                                    'Council District','Census Tract',
                                    'A', 'B','C'].sum()

In [130]:
doh_grades

ZIPCODE    Latitude   Longitude  Community Board  Council District  \
CAMIS                                                                           
30075445   83696.0  326.785850 -590.847775           1688.0             104.0   
30112340   56125.0  203.313258 -369.810407           1545.0             200.0   
30191841   50095.0  203.836629 -369.921552            520.0              15.0   
40356018   44896.0  162.319681 -295.928361           1252.0             188.0   
40356483  101106.0  365.581004 -665.162905           2862.0             414.0   
...            ...         ...         ...              ...               ...   
50115169   10002.0   40.714841  -73.991700            103.0               1.0   
50116155   33699.0  122.051758 -221.752145            948.0             123.0   
50117350   50090.0  203.762660 -369.947575            525.0              15.0   
50117434   79604.0  285.289075 -517.196049           2821.0             175.0   
50117959   30087.0  122.371566 -221.827553            333.0              24.0   

          Census Tract  A  B  C  
CAMIS                            
30075445      201600.0  1  7  0  
30112340      163500.0  5  0  0  
30191841       69500.0  5  0  0  
40356018      139200.0  4  0  0  
40356483      630000.0  5  4  0  
...                ... .. .. ..  
50115169        1600.0  1  0  0  
50116155      111900.0  3  0  0  
50117350       54500.0  5  0  0  
50117434      228900.0  0  0  7  
50117959       51000.0  3  0  0  

[19790 rows x 9 columns]

In [133]:
doh_grades[doh_grades['B'] > 0]

ZIPCODE    Latitude    Longitude  Community Board  \
CAMIS                                                          
30075445   83696.0  326.785850  -590.847775           1688.0   
40356483  101106.0  365.581004  -665.162905           2862.0   
40362264   70175.0  285.548344  -517.773967            749.0   
40362274   70084.0  285.080205  -517.982347            714.0   
40362715  230115.0  936.235858 -1702.198016           2323.0   
...            ...         ...          ...              ...   
50111805   30042.0  122.195939  -222.011203            306.0   
50113198   40008.0  162.854183  -295.984590            412.0   
50113951   41820.0  163.249463  -295.613108            808.0   
50113984   30081.0  122.429129  -221.880270            327.0   
50114370   33639.0  121.993404  -221.794553            927.0   

          Council District  Census Tract  A   B  C  
CAMIS                                               
30075445             104.0      201600.0  1   7  0  
40356483             414.0      630000.0  5   4  0  
40362264              42.0      126700.0  3   4  0  
40362274               7.0       38507.0  1   6  0  
40362715              23.0       16100.0  7  11  5  
...                    ...           ... ..  .. ..  
50111805               9.0       20100.0  0   3  0  
50113198               4.0        3200.0  0   4  0  
50113951              32.0       33200.0  0   4  0  
50113984              21.0       60900.0  0   3  0  
50114370             123.0      106500.0  0   3  0  

[9977 rows x 9 columns]

Of the 19,792 unique restaurants, 9,978 failed an initial cycle inspection at least once.

In [132]:
doh_grades[doh_grades['C'] > 0]

ZIPCODE    Latitude    Longitude  Community Board  \
CAMIS                                                          
40360076  112180.0  406.433297  -739.781887           3120.0   
40362715  230115.0  936.235858 -1702.198016           2323.0   
40363630  200560.0  815.505622 -1479.131518           2160.0   
40364179  190475.0  775.226054 -1405.243039           2033.0   
40364362  120228.0  489.124894  -887.733038           1260.0   
...            ...         ...          ...              ...   
50110216   79604.0  285.233461  -517.186091           2828.0   
50112974   70049.0  285.004311  -518.044896            707.0   
50113139  100926.0  365.433842  -665.981428           2799.0   
50113991   60162.0  244.858259  -443.760539            654.0   
50117434   79604.0  285.289075  -517.196049           2821.0   

          Council District  Census Tract  A   B   C  
CAMIS                                                
40360076             390.0      496000.0  4   0   6  
40362715              23.0       16100.0  7  11   5  
40363630             100.0      280000.0  2  11   7  
40364179             133.0      366700.0  3   5  11  
40364362              48.0      124800.0  3   3   6  
...                    ...           ... ..  ..  ..  
50110216             147.0      188307.0  0   0   7  
50112974               7.0       14700.0  0   0   7  
50113139             423.0      259200.0  0   0   9  
50113991              42.0      121800.0  0   0   6  
50117434             175.0      228900.0  0   0   7  

[5648 rows x 9 columns]

Of the 19,792 unique restaurants, 5,648 severly failed an initial cycle inspection at least once and are at risk of being closed by the DOHMH.

In [134]:
doh.duplicated().sum()

11502

In [135]:
doh.drop_duplicates(inplace=True)

In [136]:
doh.shape

(174712, 29)

In [137]:
doh['SCORE'].isna().sum()

0

In [ ]:
doh['SCORE'].hist(bins=113, figsize=(12,8));

In [ ]:
doh['SCORE'].value_counts()

In [ ]:
doh['SCORE'].mean()

In [ ]:
doh['SCORE'].median()

In [ ]:
doh['SCORE'].mode()

In [ ]:
doh['GRADE'].isna().sum()

In [ ]:
doh['GRADE'].value_counts()

In [ ]:
doh.info()

In [ ]:
doh['INSPECTION DATE'] =  pd.to_datetime(doh['INSPECTION DATE'])

In [ ]:
# doh.pivot(columns=['CAMIS','INSPECTION DATE'],values='SCORE')

The New York Health Department inspects the approximately 27,000 restaurants within the city to monitor their compliance with food safety regulations. Inspectors observe how food is prepared, served and stored and whether restaurant workers are practicing good hygiene. They check food temperatures, equipment maintenance and pest control measures.

Calling API

In [ ]:
dohmh_df['score'].hist(bins='auto', figsize=(12,8));

In [ ]:
# # Take a look at duplicated records
# duplicates = df1.duplicated(subset=['camis'], keep=False)
# df1.loc[duplicates.loc[duplicates==True].index].sort_values(by='camis')

## Yelp API

In [8]:
with open('/Users/Rob/.secret/yelp_api.json') as f:
    creds = json.load(f)

In [9]:
creds.keys()

dict_keys(['api_key'])

In [10]:
#Business Search      
# url = 'https://api.yelp.com/v3/businesses/search'

#Business Match       
# url = 'https://api.yelp.com/v3/businesses/matches'

#Phone Search         
url = 'https://api.yelp.com/v3/businesses/search/phone'

#Business Details     
# url = 'https://api.yelp.com/v3/businesses/{id}'

#Business Reviews     
# url = 'https://api.yelp.com/v3/businesses/{id}/reviews'

In [ ]:
# #Business Search  
# headers = {
#     'Authorization': 'Bearer ' + creds['api_key']
#           }
# location = 'New York NY'
# SEARCH_LIMIT = 50

# url_params = {
#     'limit': SEARCH_LIMIT,
#     'location' : location.replace(' ','+'),
#     'offset': 0
#              }


# response = requests.get(url, headers=headers, params=url_params)
# print(response.status_code)

In [140]:
phones = list(top_doh['PHONE'])
phones

['+12129569488',
 '+17187087600',
 '+12128882288',
 '+12128882288',
 '+17182719222']

In [73]:
phone_numbers = set(doh['PHONE'])

In [74]:
len(phone_numbers)

18938

In [154]:
phone_numbers = list(phone_numbers)

In [155]:
phone_numbers1 = phone_numbers[1:1000]

In [199]:
phone_numbers2 = phone_numbers[1000:2000]

In [235]:
phone_numbers3 = phone_numbers[2000:2500]

In [236]:
phone_numbers4 = phone_numbers[2500:3500]

In [254]:
phone_numbers5 = phone_numbers[3500:5000]

In [298]:
phone_numbers6 = phone_numbers[5000:6000]

In [306]:
phone_numbers7 = phone_numbers[6000:7500]

In [307]:
# Initialize empty list
biz_list7 = []

In [308]:
#Phone Search 
for number in phone_numbers7:
    url = 'https://api.yelp.com/v3/businesses/search/phone'
    headers = {'Authorization': 'Bearer ' + creds['api_key']}
    url_params = {'phone': number}
    response = requests.get(url, headers=headers, params=url_params)
    response_json = response.json()
    biz_list7.extend(response_json.get('businesses','U'))
    

In [309]:
while 'U' in biz_list7:
    biz_list7.remove('U')

In [310]:
biz_list7_df = pd.DataFrame(biz_list7)
biz_list7_df.to_csv('yelp_phone7.csv',index=False)
df = pd.read_csv('yelp_phone7.csv')
df

id                                           alias  \
0     WCiGrGn89yBrVbqXRkejwg                        jamaican-flavors-jamaica   
1     vnf1WSV7dTmOeOkIJ46hiw                             curry-king-new-york   
2     0sFAy3tueTdZYIywI_87AA                       paris-baguette-flushing-4   
3     -OplSkpDk33uikS5fO-JQw  addictive-wine-and-tapas-bar-jackson-heights-4   
4     ePZjA9S2xD6ZKpTfaH3Irg                               gong-cha-brooklyn   
...                      ...                                             ...   
1269  TynT1ykBkQhXd6nuTK6JUQ                         rego-bagels-rego-park-2   
1270  mUl0wtO7tPkBocuJqZ-7Dw              dun-huang-grand-central-new-york-4   
1271  iAmvX6GjUo7hA4ndAGhg-g        tenny-and-betsy-tea-and-pop-ups-new-york   
1272  QmrIDsmapdURUZhSAQwG6A      popeyes-louisiana-kitchen-long-island-city   
1273  REEaNSEUBROOQyrWkANqJQ                            rays-smoothies-bronx   

                                 name  \
0                    Jamaican Flavors   
1                          Curry King   
2                      Paris Baguette   
3        Addictive Wine and Tapas Bar   
4                            Gong Cha   
...                               ...   
1269                      Rego Bagels   
1270       Dun Huang -  Grand Central   
1271  Tenny & Betsy - Tea and Pop-Ups   
1272        Popeyes Louisiana Kitchen   
1273                  Ray's Smoothies   

                                                                 image_url  \
0     https://s3-media0.fl.yelpcdn.com/bphoto/GdYOODcwJLYZGd_C231qcw/o.jpg   
1     https://s3-media0.fl.yelpcdn.com/bphoto/dN8o-aaavsnCvjgvZDXhsw/o.jpg   
2     https://s3-media0.fl.yelpcdn.com/bphoto/M3prfsZvyyiRVsWfNm3_yg/o.jpg   
3     https://s3-media0.fl.yelpcdn.com/bphoto/n1_yRJ1iQEuXFaerOTy3-Q/o.jpg   
4     https://s3-media0.fl.yelpcdn.com/bphoto/4lfIopMiXui_BOcSdGwTbQ/o.jpg   
...                                                                    ...   
1269  https://s3-media0.fl.yelpcdn.com/bphoto/kwLN-ldWtcuj5Y7YfXafqA/o.jpg   
1270  https://s3-media0.fl.yelpcdn.com/bphoto/TmMr0NikR2kI8TXBHZ7qJw/o.jpg   
1271  https://s3-media0.fl.yelpcdn.com/bphoto/QE8EA1Vk4tu8wDYneNsMYQ/o.jpg   
1272  https://s3-media0.fl.yelpcdn.com/bphoto/4GUkKzWSIFXdB7KMb3xrBA/o.jpg   
1273  https://s3-media0.fl.yelpcdn.com/bphoto/xh8afqIaHQF9Xm6LuHlG1w/o.jpg   

      is_closed  \
0         False   
1         False   
2         False   
3         False   
4         False   
...         ...   
1269      False   
1270      False   
1271       True   
1272      False   
1273      False   

                                                                                                      url  \
0     https://www.yelp.com/biz/jamaican-flavors-jamaica?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_cam...   
1     https://www.yelp.com/biz/curry-king-new-york?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_campaign...   
2     https://www.yelp.com/biz/paris-baguette-flushing-4?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_ca...   
3     https://www.yelp.com/biz/addictive-wine-and-tapas-bar-jackson-heights-4?adjust_creative=82uXkAt1...   
4     https://www.yelp.com/biz/gong-cha-brooklyn?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_campaign=y...   
...                                                                                                   ...   
1269  https://www.yelp.com/biz/rego-bagels-rego-park-2?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_camp...   
1270  https://www.yelp.com/biz/dun-huang-grand-central-new-york-4?adjust_creative=82uXkAt1Tiw7u9_h33zr...   
1271  https://www.yelp.com/biz/tenny-and-betsy-tea-and-pop-ups-new-york?adjust_creative=82uXkAt1Tiw7u9...   
1272  https://www.yelp.com/biz/popeyes-louisiana-kitchen-long-island-city?adjust_creative=82uXkAt1Tiw7...   
1273  https://www.yelp.com/biz/rays-smoothies-bronx?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_campaig...   

      review_count  \
0              111   
1               68   
2              194   
3        

In [221]:
df_1 = pd.read_csv('yelp_phone1.csv')

In [223]:
df_2 = pd.read_csv('yelp_phone2.csv')

In [229]:
df_3 = pd.read_csv('yelp_phone3.csv')

In [230]:
yelp_df = pd.concat([df_1,df_2,df_3],ignore_index=True)

In [231]:
yelp_df

id                                  alias  \
0     fEsgUESZxOQtd4YEmAUuow            ii-caffe-latte-2-new-york-4   
1     zm_ngfMxyU6d85AaTIgI7Q  chloes-restaurant-and-lounge-brooklyn   
2     LbGAKVX3cwZj4fUC2XOLxg                  jahns-jackson-heights   
3     i8ejDDR4COtukAAA1Ls5fw                       soba-ya-new-york   
4     YV0VaJgV2zotzI7XNUaMCQ                 mcdonalds-new-york-100   
...                      ...                                    ...   
2159  aIRhO0WZ3KprUphL3oYsBg    applebees-grill-bar-staten-island-4   
2160  -wGvjd9NmPqrT0tm_pq1QQ                  the-black-whale-bronx   
2161  fCxQXdLdS1HLHk7bQOsy3A                       kajitsu-new-york   
2162  rn6SmSqAYx2kLFIAxytMMQ                     kajitsu-new-york-3   
2163  IjRxYyr41Yl1fGhIrss1vg                        kokage-new-york   

                             name  \
0                II Caffe Latte 2   
1     Chloe's Restaurant & Lounge   
2                          Jahn's   
3                         Soba-ya   
4                      McDonald's   
...                           ...   
2159       Applebee's Grill + Bar   
2160              The Black Whale   
2161                      Kajitsu   
2162                      Kajitsu   
2163                       Kokage   

                                                                 image_url  \
0     https://s3-media0.fl.yelpcdn.com/bphoto/rL-1bjIrEvnG6zV8ofrZpg/o.jpg   
1     https://s3-media0.fl.yelpcdn.com/bphoto/AFD_nPVpU1rhI7VMWvbltA/o.jpg   
2     https://s3-media0.fl.yelpcdn.com/bphoto/9Oa1n8Dyo_n2N9YIhJxsgw/o.jpg   
3     https://s3-media0.fl.yelpcdn.com/bphoto/LnjY25INn-aM1eAyiav8XA/o.jpg   
4     https://s3-media0.fl.yelpcdn.com/bphoto/X710tzNcQt5xzpJMVIpZHw/o.jpg   
...                                                                    ...   
2159  https://s3-media0.fl.yelpcdn.com/bphoto/oIR3BQHVVgdXBWCS7bAbJg/o.jpg   
2160  https://s3-media0.fl.yelpcdn.com/bphoto/X1ThAksqctFuD5kBszAz7w/o.jpg   
2161  https://s3-media0.fl.yelpcdn.com/bphoto/ixv3qiJXKEqSPHZRW8EMVw/o.jpg   
2162  https://s3-media0.fl.yelpcdn.com/bphoto/rbJC-rNk-wuW1AmtNUl67Q/o.jpg   
2163  https://s3-media0.fl.yelpcdn.com/bphoto/FG-id6XIpKaSJvEEwMGAwg/o.jpg   

      is_closed  \
0         False   
1         False   
2         False   
3         False   
4         False   
...         ...   
2159      False   
2160      False   
2161       True   
2162      False   
2163      False   

                                                                                                      url  \
0     https://www.yelp.com/biz/ii-caffe-latte-2-new-york-4?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_...   
1     https://www.yelp.com/biz/chloes-restaurant-and-lounge-brooklyn?adjust_creative=82uXkAt1Tiw7u9_h3...   
2     https://www.yelp.com/biz/jahns-jackson-heights?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_campai...   
3     https://www.yelp.com/biz/soba-ya-new-york?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_campaign=ye...   
4     https://www.yelp.com/biz/mcdonalds-new-york-100?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_campa...   
...                                                                                                   ...   
2159  https://www.yelp.com/biz/applebees-grill-bar-staten-island-4?adjust_creative=82uXkAt1Tiw7u9_h33z...   
2160  https://www.yelp.com/biz/the-black-whale-bronx?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_campai...   
2161  https://www.yelp.com/biz/kajitsu-new-york?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_campaign=ye...   
2162  https://www.yelp.com/biz/kajitsu-new-york-3?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_campaign=...   
2163  https://www.yelp.com/biz/kokage-new-york?adjust_creative=82uXkAt1Tiw7u9_h33zr1A&utm_campaign=yel...   

      review_count  \
0              127   
1              192   
2              172   
3             1496   
4               85   
...            ...   
2159            67   
2160           184   
2161           136   
2162           280   
2163       

In [ ]:
# def prepare_data(data_list):
#     """
#     This function takes in a list of dictionaries and prepares it
#     for analysis
#     """
    
#     # Make a new list to hold results
#     results = []
    
#     for business_data in data_list:
    
#         # Make a new dictionary to hold prepared data for this business
#         prepared_data = {}
        
#         # Extract name, review_count, rating, and price key-value pairs
#         # from business_data and add to prepared_data
#         # If a key is not present in business_data, add it to prepared_data
#         # with an associated value of None
#         for key in ("name", "review_count", "rating", "price"):
#             prepared_data[key] = business_data.get(key, None)
    
#         # Parse and add latitude and longitude columns
#         coordinates = business_data["coordinates"]
#         prepared_data["latitude"] = coordinates["latitude"]
#         prepared_data["longitude"] = coordinates["longitude"]
        
#         # Add to list if all values are present
#         if all(prepared_data.values()):
#             results.append(prepared_data)
    
#     return results
    
# # Test out function
# prepared_businesses = prepare_data(businesses)
# prepared_businesses[:5]

In [ ]:
# full_dataset = []

# for offset in get_offsets(total):
    
#     url_params['offset'] = offset
    
#     response = requests.get(url, headers=headers, params=url_params)
    
#     time.sleep(1)
    
#     response_json = response.json()
    
#     businesses = response_json.get('businesses')
# #    
#     prepared_business = prepare_data(businesses)
    
#     full_dataset.extend(prepared_businesses)

# len(full_dataset)

In [ ]:
# # Loop to retrieve all the results from the original request for all NYC businesses
# offset=0
# response = get_results('New York,NY',offset=offset)

# n_per_page = len(response['businesses'])
# total_result = response['total']

# results = response_json.get('businesses',{})

# while offset < total_result:
#     offset+=n_per_page
#     res = get_results('New York,NY',offset=offset)
#     results.extend(res['businesses'])
# results[:5]

In [ ]:
# lat = []
# long = []

# for _,business in yelp_df.iterrows():
#     lat.append(business['coordinates']['latitude'])
#     long.append(business['coordinates']['longitude'])

# yelp_df['lat'] = lat
# yelp_df['long'] = long

# Data Understanding

For this project there will be two sources and types of data used:

* Historical health and hygiene inspections recorded by New York City Department of Health and Mental Hygiene (DOHMH) public health inspectors
* User generated Yelp business ratings and reviews

## Understanding NYC DOHMH Data

This dataset contains over 330,000 records, let's explore its contents.

In [ ]:
dohmh_df['inspection_date'] =  pd.to_datetime(dohmh_inspections['inspection_date'])

In [ ]:
dohmh_df['inspection_date'].min()

In [ ]:
dohmh_df['inspection_date'].max()

Inspections in this dataset range from May 2008 up to present day.

In [ ]:
# dohmh_df.set_index('inspection_date')

In [ ]:
dohmh_df['boro'].value_counts()

In [ ]:
dohmh_df['score'].value_counts()

In [ ]:
dohmh_df['score'].isna().sum()

In [ ]:
dohmh_df['grade'].value_counts()

In [ ]:
dohmh_df['grade'].value_counts(normalize=True)

In [ ]:
dohmh_df['grade'].isna().sum()

In [ ]:
dohmh_df['critical_flag'].value_counts()

Critical violations are those most likely to contribute to foodborne illness

In [ ]:
dohmh_df['critical_flag'].value_counts(normalize=True)

In [ ]:
dohmh_df['inspection_type'].value_counts()

In [ ]:
dohmh_inspections = dohmh_df[(dohmh_df['inspection_type'] ==
                              'Cycle Inspection / Initial Inspection') | (dohmh_df['inspection_type'] ==
                              'Cycle Inspection / Re-inspection')]

In [ ]:
dohmh_inspections['critical_flag'].value_counts()

In [ ]:
# Drop records with NA critical_flag

dohmh_ohe = dohmh_inspections[dohmh_inspections['critical_flag'] != 'Not Applicable']

In [ ]:
dohmh_ohe = pd.get_dummies(dohmh_inspections,columns=['critical_flag'])

In [ ]:
dohmh_ohe.head(2)

In [ ]:
critical_flags = dohmh_ohe.groupby(['camis'])['critical_flag_Critical'].count()

In [ ]:
critical_flags[critical_flagss]

In [ ]:
critical_flags['critical_flag_Not Critical'].sum()

In [ ]:
non_critical_flags = dohmh_ohe.groupby(['camis'])

In [ ]:
dohmh_inspections.duplicated().sum()

In [ ]:
dohmh_inspections.loc[dohmh_inspections.duplicated(keep='first'),:]

In [ ]:
dohmh_inspections.drop_duplicates().shape

In [ ]:
dohmh_inspections.duplicated(subset=['camis']).sum()

In [ ]:
dohmh_inspections['camis'].nunique()

There are approximately 28,000 restaurants that have been inspected by the DOHMH. Let's check out how many have ever been flagged for a critical violation.

## Mapping with Follium